File: converted/utilities/escape-html/elide-escape-html.rb
#!/usr/bin/env ruby

# Ruby Integration Example for elide-escape-html (HTML Entity Escaper)
#
# Demonstrates calling the TypeScript escape-html implementation from Ruby
# for consistent XSS protection across microservices.

puts "=== Ruby Calling TypeScript Escape HTML ===\n"

# NOTE: Exact syntax depends on Elide's Ruby polyglot API (currently alpha)
# escape_html = Elide.require('./elide-escape-html.ts')

# Example 1: Basic HTML escaping
# dangerous = "<script>alert('XSS')</script>"
# safe = escape_html.escape(dangerous)
# puts "Dangerous: #{dangerous}"
# puts "Safe: #{safe}"
# # Output: &lt;script&gt;alert(&#39;XSS&#39;)&lt;/script&gt;

# Example 2: Rails controller with XSS protection
# class CommentsController < ApplicationController
#   def create
#     user_input = params[:comment][:body]
#
#     # Use TypeScript HTML escaper from Ruby
#     escape_html = Elide.require('./elide-escape-html.ts')
#     safe_comment = escape_html.escape(user_input)
#
#     @comment = Comment.create!(body: safe_comment)
#     render json: { comment: safe_comment, status: 'safe' }
#   end
# end

# Example 3: Sinatra app with consistent escaping
# require 'sinatra'
# ESCAPE_HTML = Elide.require('./elide-escape-html.ts')
#
# post '/api/post' do
#   user_title = params['title']
#   user_body = params['body']
#
#   # Escape using same implementation as Node.js
#   safe_title = ESCAPE_HTML.escape(user_title)
#   safe_body = ESCAPE_HTML.escape(user_body)
#
#   # Save to database (XSS-safe)
#   Post.create(title: safe_title, body: safe_body)
#
#   json status: 'success', title: safe_title
# end

# Example 4: Email template generation
# def send_notification_email(user)
#   escape_html = Elide.require('./elide-escape-html.ts')
#
#   # Escape user data for HTML email
#   safe_name = escape_html.escape(user.name)
#   safe_message = escape_html.escape(user.message)
#
#   html_body = <<-HTML
#     <html>
#       <body>
#         <h1>Hello, #{safe_name}!</h1>
#         <p>#{safe_message}</p>
#       </body>
#     </html>
#   HTML
#
#   Mailer.send(user.email, "Notification", html_body)
# end

# Example 5: Sidekiq job with HTML sanitization
# class RenderHtmlWorker
#   include Sidekiq::Worker
#
#   def perform(user_id, content)
#     escape_html = Elide.require('./elide-escape-html.ts')
#
#     # Sanitize content before rendering
#     safe_content = escape_html.escape(content)
#
#     # Generate HTML report
#     html = generate_report(safe_content)
#     save_report(user_id, html)
#   end
# end

# Example 6: API response sanitization
# require 'grape'
# class API < Grape::API
#   format :json
#
#   helpers do
#     def escape_html
#       @escape_html ||= Elide.require('./elide-escape-html.ts')
#     end
#   end
#
#   get '/users/:id' do
#     user = User.find(params[:id])
#
#     # Escape all user-generated fields
#     {
#       name: escape_html.escape(user.name),
#       bio: escape_html.escape(user.bio),
#       website: escape_html.escape(user.website)
#     }
#   end
# end

puts "Real-world use case:"
puts "- Ruby Rails app receives user input"
puts "- Uses same HTML escaper as Node.js frontend"
puts "- Guarantees identical XSS protection across all services"
puts ""

puts "Example: Unified XSS Protection"
puts "┌─────────────────────────────────────┐"
puts "│   Elide Escape HTML (TypeScript)   │"
puts "│   conversions/escape-html/*.ts     │"
puts "└─────────────────────────────────────┘"
puts "         ↓           ↓"
puts "    ┌────────┐  ┌────────┐"
puts "    │ Node.js│  │  Ruby  │"
puts "    │  SSR   │  │  Rails │"
puts "    └────────┘  └────────┘"
puts "     escape(input)   escape(input)"
puts "     = &lt;safe&gt;   = &lt;safe&gt;"
puts "     ✓ Consistent protection!"
puts ""

puts "Problem Solved:"
puts "Before: ERB::Util.html_escape vs Node.js escaping = inconsistencies"
puts "After: One Elide implementation = perfect XSS prevention"
puts ""

puts "Security Example:"
puts "  # User input"
puts "  user_input = '<img src=x onerror=alert(1)>'"
puts "  "
puts "  # Escape using shared implementation"
puts "  safe = escape_html.escape(user_input)"
puts "  # Result: &lt;img src=x onerror=alert(1)&gt;"
puts "  "
puts "  # Same escaping in Node.js, Python, Ruby, Java!"
puts "  # ✓ Unified security standard"

File: converted/utilities/escape-html/elide-escape-html.py
#!/usr/bin/env python3
"""
Python Integration Example for elide-escape-html (HTML Entity Escaper)

Demonstrates calling the TypeScript escape-html implementation from Python
for consistent XSS protection and HTML escaping across all services.

Benefits:
- One HTML escaper shared across TypeScript and Python
- Consistent XSS prevention across web services
- No Python-specific escaping quirks or library differences
"""

# NOTE: Exact syntax depends on Elide's Python polyglot API (currently alpha)
# from elide import require
# escape_html = require('./elide-escape-html.ts')

print("=== Python Calling TypeScript Escape HTML ===\n")

# Example 1: Basic HTML escaping
# dangerous = "<script>alert('XSS')</script>"
# safe = escape_html(dangerous)
# print(f"Dangerous: {dangerous}")
# print(f"Safe: {safe}")
# # Output: &lt;script&gt;alert(&#39;XSS&#39;)&lt;/script&gt;

# Example 2: User comment sanitization
# comment = "I love <3 this & that's \"great\"!"
# safe_comment = escape_html(comment)
# print(f"\nComment: {comment}")
# print(f"Safe: {safe_comment}")
# # Output: I love &lt;3 this &amp; that&#39;s &quot;great&quot;!

# Example 3: Flask application with XSS protection
# from flask import Flask, request, jsonify
# app = Flask(__name__)
#
# # Use TypeScript HTML escaper from Python
# @app.route('/api/comment', methods=['POST'])
# def post_comment():
#     user_input = request.json.get('comment', '')
#
#     # Escape using shared TypeScript implementation
#     safe_comment = escape_html.escape(user_input)
#
#     # Save to database (now XSS-safe)
#     save_comment(safe_comment)
#
#     return jsonify({'comment': safe_comment, 'status': 'safe'})

# Example 4: Django template rendering
# # views.py
# from django.shortcuts import render
# from django.utils.safestring import mark_safe
#
# def user_profile(request):
#     user_bio = request.POST.get('bio', '')
#
#     # Escape using same implementation as Node.js service
#     safe_bio = escape_html.escape(user_bio)
#
#     context = {
#         'bio': mark_safe(safe_bio)  # Already escaped
#     }
#     return render(request, 'profile.html', context)

# Example 5: Email HTML generation
# def send_welcome_email(user_name, user_email):
#     # Escape user data to prevent HTML injection in emails
#     safe_name = escape_html.escape(user_name)
#     safe_email = escape_html.escape(user_email)
#
#     html_body = f"""
#     <html>
#         <body>
#             <h1>Welcome, {safe_name}!</h1>
#             <p>Your email: {safe_email}</p>
#         </body>
#     </html>
#     """
#
#     send_email(user_email, "Welcome", html_body)

# Example 6: API response sanitization
# from fastapi import FastAPI
# app = FastAPI()
#
# @app.get("/user/{user_id}")
# async def get_user(user_id: int):
#     user = get_user_from_db(user_id)
#
#     # Escape all user-generated content
#     return {
#         "name": escape_html.escape(user.name),
#         "bio": escape_html.escape(user.bio),
#         "website": escape_html.escape(user.website)
#     }

print("Real-world use case:")
print("- Python web service receives user input")
print("- Uses same HTML escaper as Node.js frontend")
print("- Guarantees identical XSS protection across entire stack")
print()

print("Example: Unified XSS Protection")
print("┌─────────────────────────────────────┐")
print("│   Elide Escape HTML (TypeScript)   │")
print("│   conversions/escape-html/*.ts     │")
print("└─────────────────────────────────────┘")
print("         ↓                   ↓")
print("    ┌────────┐          ┌────────┐")
print("    │ Node.js│          │ Python │")
print("    │  SSR   │          │  API   │")
print("    └────────┘          └────────┘")
print("     escapeHtml(input)   escapeHtml(input)")
print("     = &lt;safe&gt;      = &lt;safe&gt;")
print("     ✓ Consistent XSS prevention!")
print()

print("Problem Solved:")
print("Before: Node.js escapes differently than Python = XSS vulnerabilities")
print("After: One Elide implementation = unified XSS protection")
print()

print("Security Example:")
print("  # User input from form")
print("  user_input = '<script>alert(1)</script>'")
print("  ")
print("  # Escape using shared implementation")
print("  safe = escape_html(user_input)")
print("  # Result: &lt;script&gt;alert(1)&lt;/script&gt;")
print("  ")
print("  # Same escaping logic in Node.js, Python, Ruby, Java!")
print("  # One security standard = fewer vulnerabilities")
